### PR TITLE
[python] Explicit shaping for `DataFrame`

### PR DESCRIPTION
**Pull request by johnkerl**
Context: https://github.com/single-cell-data/TileDB-SOMA/issues/961

This is a prerequisite for https://github.com/single-cell-data/TileDB-SOMA/pull/1027.

Note that for `SparseNDArray` it sufficed to have a slotwise int `shape`: indexing is only by `int64` and we took the min to be 0, leaving only the max to be specified. Here, for `DataFrame`, indices can be of various types so we truly must do parameterizations in terms of domains -- min-max pairs on each dim slot.

### CODE REVIEW

File: python-spec/src/somacore/data.py [hunk 1]
```diff
@@ -58,14 +59,24 @@ def create(
...
             must define all columns, including columns to be named as index
             columns.  If the schema includes types unsupported by the SOMA
             implementation, an error will be raised.
+
         :param index_column_names: A list of column names to use as user-defined
             index columns (e.g., ``['cell_type', 'tissue_type']``).
             All named columns must exist in the schema, and at least one
             index column name is required.
+
+        :param domains: An optional sequence of min-max tuples specifying the
```
Comment (bkmartinjr) [inline python-spec/src/somacore/data.py:73]: same comments as on the `tiledbsoma` PR review - the param description is confusing.  Suggest making it consistent with the other PR

File: python-spec/src/somacore/data.py [hunk 2]
```diff
@@ -45,6 +45,7 @@ def create(
         *,
         schema: pa.Schema,
         index_column_names: Sequence[str] = (options.SOMA_JOINID,),
+        domains: Optional[Sequence[Optional[Tuple[Any, Any]]]] = None,
```
Comment (bkmartinjr) [inline python-spec/src/somacore/data.py:48]: the thing missing from this is access to the domain/extent on an existing DataFrame.

We used to have `shape` (it still exists on NDArray), and I thought we intended to bring that back once the DataFrame had added this first-class concept.

So, where is the `shape` property equivalent?   And what should it be called?   (`domain_extent`?)

CC: @thetorpedodog

**Review by bkmartinjr — DISMISSED** [review]
the major open issue IMHO is the addition of the domain, without a property to access it on existing DataFrames.  We need the accessor (much like NDArray has `shape` on both `create` and as a read-only property on existing NDArrays).

File: python-spec/src/somacore/data.py [hunk 3]
```diff
@@ -170,6 +185,14 @@ def index_column_names(self) -> Tuple[str, ...]:
         """
         raise NotImplementedError()
 
+    @property
+    def domains(self) -> Tuple[Tuple[Any, Any], ...]:
```
Comment (johnkerl) [inline python-spec/src/somacore/data.py:189]: I'm now thinking this should be singular `domain` in line with tiledb usage.

If everyone's agreed I'll of course make the same change in the impl PR https://github.com/single-cell-data/TileDB-SOMA/pull/1027.

File: python-spec/src/somacore/data.py [hunk 4]
```diff
@@ -170,6 +185,14 @@ def index_column_names(self) -> Tuple[str, ...]:
         """
         raise NotImplementedError()
 
+    @property
```
Comment (thetorpedodog) [inline python-spec/src/somacore/data.py:188]: don’t forget to add an `@abc.abstractmethod` here.

**Review by thetorpedodog — COMMENTED** [review]
I have no objections here (apart from the one note below); if Brian is happy with it I am.

File: python-spec/src/somacore/data.py (hunk 3 above)
Comment (bkmartinjr) [reply python-spec/src/somacore/data.py]: I prefer the singular as well.  @thetorpedodog ?

**Comment by bkmartinjr** [comment]
> if Brian is happy with it I am

I don't think Brian is on the review - add him?

(joking - I assume this was a typo)

**Review by bkmartinjr — APPROVED** [review]

**Pull request merged**